### PR TITLE
Foxy update documentation

### PR DESCRIPTION
**Pull request by carlossvg**
This is a draft to update the documentation as listed in https://github.com/ros2-realtime-demo/pendulum/issues/35.

## Checklist

- [ ] Review grammar and typos
- [ ] Review install instructions
- [ ] Review links

### CODE REVIEW

**Comment by carlossvg** [comment]
@LanderU I decided to create a specific document with the install instructions and provide different installation methods. What do you think?

**Comment by carlossvg** [comment]
@LanderU I propose to merge a preliminary version of this branch with the new structure and the basic parts (e.g: ADE install, run instructions) and leave the pending sections with a TODO.

File: docs/installation.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# Install instructions
+
+## ADE based installation
+
+TODO: describe how the ADE image is structured and which features already includes (ROS 2traces
```
Comment (carlossvg) [inline docs/installation.md:5]: @LanderU Could you add a brief description of the ADE image?

Comment (LanderU) [reply docs/installation.md:5]: Yes, let me review what you've done and I'll add the remaining stuff.

File: README.md [hunk 2]
```diff
@@ -2,120 +2,100 @@
...
+**Foxy branch is under active development. The current status can be checked here: 
+https://github.com/ros2-realtime-demo/pendulum/issues/35**
 
-### Build status
+## Build status
 
-| foxy | eloquent |  dashing |
-|--------- | -------- | ------- |
-| ![Build Status](https://github.com/ros2-realtime-demo/pendulum/workflows/build/badge.svg?branch=foxy) | ![Build Status](https://github.com/ros2-realtime-demo/pendulum/workflows/build/badge.svg?branch=eloquent) | ![Build Status](https://github.com/ros2-realtime-demo/pendulum/workflows/build/badge.svg?branch=dashing) |
+![Build Status](https://github.com/ros2-realtime-demo/pendulum/workflows/build/badge.svg?branch=foxy) [![License](https://img.shields.io/badge/license-Apache%202-blue)]() ![GitHub issues](https://img.shields.io/github/issues/ros2-realtime-demo/pendulum)
```
Comment (LanderU) [inline README.md:14]: Why this change?

File: README.md [hunk 3]
```diff
@@ -2,120 +2,100 @@
...
+## Install instructions
 
-In order to get the project ready, first you need to install [ROS2](https://index.ros.org/doc/ros2/Installation/Eloquent/) and [rosdep](http://wiki.ros.org/rosdep).
+ Several ways to install the pendulum demo project are provided depending on the user needs. 
+ 
+- [ADE based installation](docs/installation.md#ade#based#installation): This provides a 
+pre-configured Docker image with everything ready to be used. This is the recommended option to
+ quickly test the demo. It is also useful to have development environment to add new features.
+- [Installation from source](docs/installation.md#installation#from#source): provides instructions
+ to build all the required ROS 2 packages from source
```
Comment (LanderU) [inline README.md]:
```suggestion
 to build all the required ROS 2 packages from source.
```

File: README.md [hunk 4]
```diff
@@ -2,120 +2,100 @@
...
+- [Crosscompilation](docs/installation.md#crosscompile): provides instructions to
+ crosscompile the demo for an embedded device
```
Comment (LanderU) [inline README.md]:
```suggestion
 crosscompile the demo for an embedded device.
```

File: README.md [hunk 5]
```diff
@@ -2,120 +2,100 @@
...
+- [Pre-configured Raspberry PI image](docs/installation.md#raspberry#pi#image): provides a
+ Raspberry PI with everything configured in a Linux Real-time
```
Comment (LanderU) [inline README.md]:
```suggestion
 Raspberry PI with everything configured in a Linux Real-time (RT-PREEMPT patch).
```

File: README.md [hunk 6]
```diff
@@ -2,120 +2,100 @@
...
-```
+### Source your workspace
 
-#### How to run the demo
+If installed as an ADE image:
 
-Terminal 1:
 ```bash
-ros2 launch pendulum_bringup pendulum_bringup.launch.py
+source /opt/ros/foxy/setup.bash
```
Comment (LanderU) [inline README.md:60]: This will change with the upcoming work.

File: README.md [hunk 7]
```diff
@@ -2,120 +2,100 @@
...
 
-A detailed tutorial for real-time testing can be found [here](docs/real_time_tutorial.md).
+TODO
 
-#### How to contribute
+## How to contribute
 
-Make a fork using github, and make your changes. This project uses the same guidelines as all the officials ROS2 packages, make sure that your code is compliant with cppcheck, uncrustify...
+Make a fork using github, and make your changes. This project uses the same guidelines as all the
+ officials ROS2 packages, make sure that your code is compliant with cppcheck, uncrustify...
```
Comment (LanderU) [inline README.md]:
```suggestion
 officials ROS 2 packages, make sure that your code is compliant with cppcheck, uncrustify...
```

File: docs/design.md [hunk 8]
```diff
@@ -2,36 +2,35 @@
 
 ### Purpose
 
-This is a rework of the existing pendulum demo in ros2/demos. In this project we want to create a new version that updates some the new features of ROS 2 such as compositions, lifecycle, new QoS and real-time support.
+This is a rework of the existing pendulum demo in ros2/demos. The aim of this project is to
```
Comment (LanderU) [inline docs/design.md]:
```suggestion
This is a rework of the existing pendulum demo in [ros2/demos](https://github.com/ros2/demos). The aim of this project is to
```

File: docs/installation.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+```bash
+$ which ade
+/path/to/ade
+$ ade --version
+<version>
+```
+
+### Setup ADE home and project checkout
+ADE needs a directory on the host machine which is mounted as the user's home directory within
```
Comment (LanderU) [inline docs/installation.md:32]:
```suggestion

ADE needs a directory on the host machine which is mounted as the user's home directory within
```

File: docs/installation.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+$ cd adehome
+$ touch .adehome
+```
+
+For ADE to function, it must be properly configured. The pendulum demo provides an `.aderc` file
+ which is expected to exist in the current working directory, or in any parent directory. 
+
+```
+$ cd adehome
+$ git clone git@github.com:ros2-realtime-demo/pendulum.git
```
Comment (LanderU) [inline docs/installation.md:54]: To clone this way, the user needs to have ssh key added to their github. We can avoid issues with that cloning the repo like: `git clone https://github.com/ros2-realtime-demo/pendulum`

File: docs/installation.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+ade$ source /opt/ros/foxy/setup.bash
+ade$ cd pendulum
+ade$ colcon build
+ade$ colcon test
+ade$ colcon test-result
+```
+
+## Installation from source
+
+TODO
```
Comment (LanderU) [inline docs/installation.md]: I need to complete this part.

File: docs/installation.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+ade$ colcon test-result
+```
+
+## Installation from source
+
+TODO 
+
+## Crosscompile
+ 
+TODO
```
Comment (LanderU) [inline docs/installation.md:71]: I need to complete this part.

File: docs/installation.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+ 
+## Raspberry PI image
+
+TODO
```
Comment (LanderU) [inline docs/installation.md]: I need to complete this part.

File: docs/real_time_tutorial.md [hunk 14]
```diff
@@ -2,101 +2,42 @@
...
-* https://index.ros.org/doc/ros2/Tutorials/Allocator-Template-Tutorial/
+These are the options that allows to configure the process real-time settings:
+* `priority`: changes the process priority and set a real-time FIFO priority. This will set the
+ priority of the thread where the ROS 2 executor is running.
+* `cpu-affinity`: binds the application to a specific CPU core by setting a CPU mask. For
+ example, to bind the process or thread to CPU 3 (or 2 starting from 0) we set the CPU mask to 4
+  (100 in binary).
+* `lock-memory`: pre-faults memory until no more page faults are seen. This usually allocated a 
+ high amount of memory so make sure there is enough memory in the system.
+* `lock-memory`: specifies the amount of memory we want to pre-allocate. For example
```
Comment (LanderU) [inline docs/real_time_tutorial.md]:
```suggestion
* `lock-memory-size`: specifies the amount of memory we want to pre-allocate. For example 
```

File: docs/real_time_tutorial.md [hunk 15]
```diff
@@ -2,101 +2,42 @@
...
-ros2 run pendulum_demo pendulum_demo --lock-memory-size 50000
-```
-
-If we get the following error message this means we don't have permissions or that we can't allocate the requested amount of memory.
-
-```
-mlockall failed: Cannot allocate memory
-Couldn't lock  virtual memory.
+```bash
+ros2 run pendulum_demo pendulum_demo --priority 80 --cpu-affinity:=4 --lock-memory-size 100 --config-child-threads True
```
Comment (LanderU) [inline docs/real_time_tutorial.md:38]:
```suggestion
ros2 run pendulum_demo pendulum_demo --priority 80 --cpu-affinity:=4 --lock-memory-size 100 --config-child-threads:=True 
```

File: README.md (hunk 2 above)
Comment (carlossvg) [reply README.md:14]: I thought that makes more sense to only show the build status for the current branch. For every release, we will be adding a new badge, also if we stop supporting an old branch and it stops passing I would prefer not to show it. I don't have a strong opinion on this, change this as you think is best.

File: README.md (hunk 6 above)
Comment (carlossvg) [reply README.md:60]: Right, if we merge this first, the instructions should be consistent with the current status. The instructions should be changed accordingly with the upcoming work.

File: docs/real_time_tutorial.md (hunk 15 above)
Comment (carlossvg) [reply docs/real_time_tutorial.md:38]: That only applies to the launch arguments, this is for the executable. Anyhow, we should validate all the commands before approving the PR.

File: README.md (hunk 6 above)
Comment (LanderU) [reply README.md:60]: I'll include a file on the volume that source it.

File: README.md (hunk 2 above)
Comment (LanderU) [reply README.md:14]: Fair enough.

File: docs/installation.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+ade$ colcon build
+ade$ colcon test
+ade$ colcon test-result
+```
+
+## Installation from source
+
+TODO
+
+## Cross compile
```
Comment (LanderU) [inline docs/installation.md:69]: @carlossvg, could you please take a look at this?

File: docs/installation.md [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+* Python >= 3.5
+
+```bash
+sudo apt-get install qemu-user-static
+```
+
+For more information, checkout: https://github.com/ros-tooling/cross_compile#installation
+
+```bash
+mkdir ~/xcompile_ws/src
```
Comment (carlossvg) [inline docs/installation.md]: mkdir -p ~/xcompile_ws/src

File: docs/installation.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+cd ~/xcompile_ws/src
+git clone https://github.com/ros2-realtime-demo/pendulum -b foxy
```
Comment (carlossvg) [inline docs/installation.md]: I would use ROS_DISTRO:

```
ROS_DISTRO=foxy
git clone https://github.com/ros2-realtime-demo/pendulum -b $ROS_DISTRO
...
ros_cross_compile xcompile_ws/ --arch aarch64 --os ubuntu --rosdistro $ROS_DISTRO
```

File: docs/installation.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+pip3 install ros_cross_compile
+cd ~
+ros_cross_compile xcompile_ws/ --arch aarch64 --os ubuntu --rosdistro foxy
```
Comment (carlossvg) [inline docs/installation.md]: I would move to requirements, just after installing quemu

File: docs/installation.md [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+In order to cross compile the project to use it on aarch64 CPU architecture, here the needed steps.
+
+### Prerequisites
+
+* [Docker](https://docs.docker.com/engine/install/ubuntu/), add current user to docker group to execute it as non-root user.
+* Python >= 3.5
+
+```bash
+sudo apt-get install qemu-user-static
+pip3 install ros_cross_compile
```
Comment (carlossvg) [inline docs/installation.md]: This required sudo in my case

File: docs/installation.md [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+
+For more information, checkout: https://github.com/ros-tooling/cross_compile#installation
+
+```bash
+mkdir -p ~/xcompile_ws/src
+cd ~/xcompile_ws/src
+export ROS_DISTRO=foxy
+git clone https://github.com/ros2-realtime-demo/pendulum -b "${ROS_DISTRO}"
+cd ~
+ros_cross_compile xcompile_ws/ --arch aarch64 --os ubuntu --rosdistro  "${ROS_DISTRO}"
```
Comment (carlossvg) [inline docs/installation.md]: This failed following the instructions with the following error:

```bash
INFO:Docker Client:+ echo 'No src/ directory found at /ws, did you remember to mount your workspace?'
INFO:Docker Client:+ exit 1
INFO:Docker Client:No src/ directory found at /ws, did you remember to mount your workspace?
```

Comment (LanderU) [reply docs/installation.md]:
```bash
ros_cross_compile /home/$USER/xcompile_ws/ --arch aarch64 --os ubuntu --rosdistro  "${ROS_DISTRO}"
```

Comment (carlossvg) [reply docs/installation.md]: Thanks, this worked for me. I was ablet o cross-compile for a RPI3 and Ubuntu 20.04.

File: docs/installation.md (hunk 16 above)
Comment (carlossvg) [reply docs/installation.md:69]: @LanderU I removed this section for the moment

Comment (LanderU) [reply docs/installation.md:69]: Perfect!

**Review by LanderU — APPROVED** [review]
LGTM! We need to work more on this
